contracts/Governance/Timelock.sol

```diff
@@ -1,123 +1,160 @@
+        require(_delay <= maximumDelay, "Timelock: Delay must not exceed maximum delay.");
+
+        delay = _delay;
 
-        emit NewDelay(delay);
+        emit NewDelaySet(delay);
     }
 
-    function acceptAdmin() external {
-        require(msg.sender == pendingAdmin, "Timelock::acceptAdmin: Call must come from pendingAdmin.");
-        admin = msg.sender;
-        pendingAdmin = address(0);
+    function queueTransactionsBatch(Transaction[] memory transactions, uint256 eta) external onlyProposerOrOwner returns (bytes32) {
+        //todo ag hmm... do we really want it this way? we definitely want 0 delay on local deployment
+        require(eta >= block.timestamp + delay, "Timelock: Estimated execution time must satisfy delay.");
+        require(eta <= block.timestamp + delay + gracePeriod, "Timelock: Estimated execution time must satisfy delay and grace period.");
 
-        emit NewAdmin(admin);
+        bytes32 txParamsHash = keccak256(abi.encode(transactions, eta));
+        queuedTransactions[txParamsHash] = TrnasactionStatus.Queued;
+
+        emit QueuedTransactionsBatch(txParamsHash, transactions.length, eta);
+        return txParamsHash;
     }
 
-    function setPendingAdmin(address pendingAdmin_) external {
-        require(msg.sender == address(this), "Timelock::setPendingAdmin: Call must come from Timelock.");
-        pendingAdmin = pendingAdmin_;
+    function cancelTransactionsBatch(bytes32 txParamsHash) external onlyProposerOrOwner {
+        require(queuedTransactions[txParamsHash] != TrnasactionStatus.NonExistent, "Timelock: transaction is not queued");
+
+        delete queuedTransactions[txParamsHash];
 
-        emit NewPendingAdmin(pendingAdmin);
+        emit CancelledTransactionsBatch(txParamsHash);
     }
 
-    function queueTransaction(
-        address target,
-        uint256 value,
-        string memory signature,
-        bytes memory data,
-        uint256 eta
-    ) external returns (bytes32) {
-        require(msg.sender == admin, "Timelock::queueTransaction: Call must come from admin.");
-        require(eta >= getBlockTimestamp() + delay, "Timelock::queueTransaction: Estimated execution block must satisfy delay.");
-
-        bytes32 txHash = keccak256(abi.encode(target, value, signature, data, eta));
-        queuedTransactions[txHash] = true;
-
-        emit QueueTransaction(txHash, target, value, signature, data, eta);
-        return txHash;
+    function approveTransactionsBatch(bytes32 txParamsHash) external onlyOwner {
+        _approveTransactionsBatchInternal(txParamsHash);
     }
 
-    function cancelTransaction(
-        address target,
-        uint256 value,
-        string memory signature,
-        bytes memory data,
-        uint256 eta
-    ) external {
-        require(msg.sender == admin, "Timelock::cancelTransaction: Call must come from admin.");
+    function executeTransactionsBatch(Transaction[] memory transactions, uint256 eta) external payable onlyProposerOrOwner {
```

Comment:
Why are we giving the proposer the option to execute the transaction?
Shouldn't that be reserved only for the multisig? The way I see it the proposer should only have the power to propose transactions (hence its name).
Otherwise, it's not really proposer.. It would be a propoerAndExecutor